pico.py

```diff
@@ -6,43 +6,126 @@
 HOST='127.0.0.1'
 PORT=5000
 
+def is_connection_closed(conn) -> bool:
+    try:
+        # this will try to read bytes without blocking and also without removing them from buffer (peek only)
+        data = conn.recv(4, socket.MSG_DONTWAIT | socket.MSG_PEEK)
+        if len(data) == 0:
+            return True
+    except BlockingIOError:
+        return False  # socket is open and reading from it would block
+    except ConnectionResetError:
+        return True  # socket was closed for some other reason
+    except Exception as e:
+        return False
+    return False
+
+def is_valid_command(command):
+	command_length = len(bin(command)[2:])
+
+	if command_length < 24:
+		print(f"Invalid command: {bin(command)}")
+		return False
+	else:
+		return True
+	
+
+def get_command_header(command):
+	# move all bits except the first 4 to the right
+	# only the first 4 bits will be returned
+
+	if not is_valid_command(command):
+		return 0
+
+	return command >> len(bin(command)[2:]) - 4
```

Comment:
Gets the first 4 bits using bit shifting.

`1111000010100101 >> 12 == 1111` would mean moving everything over 12 places

This is the binary version of dividing by 10, 100, 1000, etc in decimal.

`4000 / 1000 == 4`